test_conformance/subgroups/workgroup_common_templates.h

```diff
@@ -0,0 +1,585 @@
+//
+// Copyright (c) 2020 The Khronos Group Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#ifndef WORKGROUPCOMMONTEMPLATES_H
+#define WORKGROUPCOMMONTEMPLATES_H
+
+#include "typeWrappers.h"
+#include <bitset>
+#include "CL/cl_half.h"
+
+
+typedef std::bitset<128> bs128;
+inline cl_uint4 generate_bit_mask(cl_uint subgroup_local_id,
+                                  std::string mask_type,
+                                  cl_uint max_sub_group_size)
+{
+    bs128 mask128;
+    cl_uint4 mask;
+    cl_uint pos = subgroup_local_id;
+    if (mask_type == "eq") mask128.set(pos);
+    if (mask_type == "le" || mask_type == "lt")
+    {
+        for (cl_uint i = 0; i <= pos; i++) mask128.set(i);
+        if (mask_type == "lt") mask128.reset(pos);
+    }
+    if (mask_type == "ge" || mask_type == "gt")
+    {
+        for (cl_uint i = pos; i < max_sub_group_size; i++) mask128.set(i);
+        if (mask_type == "gt") mask128.reset(pos);
+    }
+
+    // convert std::bitset<128> to uint4
+    auto const uint_mask = bs128{ static_cast<unsigned long>(-1) };
+    mask.s0 = (mask128 & uint_mask).to_ulong();
+    mask128 >>= 32;
+    mask.s1 = (mask128 & uint_mask).to_ulong();
+    mask128 >>= 32;
+    mask.s2 = (mask128 & uint_mask).to_ulong();
+    mask128 >>= 32;
+    mask.s3 = (mask128 & uint_mask).to_ulong();
+
+    return mask;
+}
+
+// DESCRIPTION :
+// Which = 0 -  sub_group_broadcast - each work_item registers it's own value.
+// All work_items in subgroup takes one value from only one (any) work_item
+// Which = 1 -  sub_group_broadcast_first - same as type 0. All work_items in
+// subgroup takes only one value from only one chosen (the smallest subgroup ID)
+// work_item Which = 2 -  sub_group_non_uniform_broadcast - same as type 0 but
+// only 4 work_items from subgroup enter the code (are active)
+
+template <typename Ty, int Which = 0> struct BC
```

Comment:
If this type is local to the file, it probably should be enclosed in an anon namespace

Reply:
this is used in many files